protocol/eth-shared-lockbox.md

```diff
@@ -0,0 +1,164 @@
+        1. For the chain being added, it triggers `addDependency` with `chainId1`, `chainId2`, … `chainIdn`.
+        2. For all existing chains, it triggers `addDependency` with `chainId` as the newly added chain.
+    2. Updates the dependency graph (mapping or array).
+    3. Emits an event for step (1).
+
+A code example for step (2) would look like this:
+
+```solidity
+// Mapping from chainId to SystemConfigInterop address
+mapping(uint256 => address) public systemConfigInterops;
+// Mapping to check if a chainId is in the dependency set
+mapping(uint256 => bool) public alreadyAdded;
+// Current dependency set list
+uint256[] public dependencySet;
+
+// Function to add a chain to the dependency set
+function addChain(uint256 chainId) external onlyOwner {
+		require(systemConfigInterops[chainId] != address(0), "DependencySetManager: Chain not registered");
+		require(!isRegistered[chainId], "DependencySetManager: Chain already in dependency set");
+
+		// For the chain being added, call addDependencies with existing chainIds
+		if (dependencySet.length > 0) {
+				for (uint256 i = 0; i < dependencySet.length; i++) {
+            ISystemConfigInterop(systemConfigInterops[chainId]).addDependency(dependencySet[i]);
+            }
+		}
+
+		// For each existing chain, call addDependencies with the new chainId
+		for (uint256 i = 0; i < dependencySet.length; i++) {
+				ISystemConfigInterop(systemConfigInterops[dependencySet[i]]).addDependency(chainId);
+    }
+    // Update the dependency set
+    dependencySet.push(chainId);
+    isRegistered[chainId] = true;
+
+    emit ChainAdded(chainId);
+}
+    
+```
+
+For `removeChain`, the process would follows the same logic.
+
+Note that, under the specified flow, the dependency set consistently maintains the form of a [complete graph](https://en.wikipedia.org/wiki/Complete_graph) at all times.
+
+### `addDependencies` and `removeDependencies` in `SystemConfigInterop` and `L1BlockInterop`
+
+The purpose of these functions is to extend the current `SystemConfigInterop` implementation by allowing multiple `chainId` values to be added in a single call by passing them as an array. There are a few ways to pass these values through `setConfig`: one is making a single call and triggering the `TransactionDeposited` event, or by looping through the current `addDependency`/`removeDependency` functions.
+
+With this optimization, the `DependencySetManager` can simplify the number of calls, as follows:
+
+```solidity
+function addChain(uint256 chainId) external onlyOwner {
+  // ...
+	if (dependencySet.length > 0) {
+				// If we can input an array chainId values thorugh addDependencies function
+				ISystemConfigInterop(newChainSystemConfig).addDependencies(dependencySet);
+	}		
+	// ...
+```
+
+### `SharedLockbox` implementation
+
+A minimal set of functions should include:
+
+- `lockETH`: Accepts ETH from the `depositTransaction` originating from a valid `OptimismPortal2`.
```

Comment:
Once we define the migration approach, we will try to apply the code changes needed and check if it reach the size limit. If we are able to avoid hitting the limit, we will leave it like that and apply the changes to the portal. If we hit the size limit, we will need to revisit and tackle the possible portal decoupling.